webapp/src/auth.ts

```diff
@@ -159,46 +155,92 @@ export function identityProviders(): pxt.AppCloudProvider[] {
         .filter(prov => prov.identity);
 }
 
+export function hasIdentity(): boolean {
+    return identityProviders().length > 0;
+}
+
+export function loggedIn(): boolean {
+    if (!hasIdentity()) return false;
+    const state = getState();
+    return !!state.user?.id
+}
+
+export function profileNeedsSetup(): boolean {
+    const state = getState();
+    return loggedIn() && !state.user.username;
+}
+
+export async function updateUserProfile(opts: {
+    username?: string,
+    avatarUrl?: string
+}) {
+    if (!loggedIn()) return;
+    const result = await apiAsync<UserProfile>('/api/user/profile', {
+        username: opts.username,
+        avatarUrl: opts.avatarUrl
+    } as UserProfile);
+    if (result.success) {
+        // Set user profile from returned value
+        setUser(result.resp);
+    }
+}
+
 /**
  * Private functions
  */
 
-const redirBase = () => `${window.location.protocol}//${window.location.host}`;
-
 async function fetchUserAsync() {
     const state = getState();
-    if (state.user) return;
 
-    const resp = await apiAsync('user/profile');
-    const user: UserProfile = resp.json;
-
-    setUser(user);
-}
-
-function mkLoginStateKey(id: string): string {
-    return `auth.login.state:${id}`;
-}
+    // We already have a user, no need to get it again.
+    if (state.user) return;
 
-function authEnabled(): boolean {
-    return identityProviders().length > 0;
+    const result = await apiAsync('/api/user/profile');
+    if (result.success) {
+        setUser(result.resp);
+    }
 }
 
 function idpEnabled(idp: pxt.IdentityProviderId): boolean {
     return identityProviders().filter(prov => prov.id === idp).length > 0;
 }
 
-function setUser(user: UserProfile) {
+function setUser(user: UserProfile, invalidate: boolean = true) {
     state_.user = user;
-    data.invalidate(AUTH_USER);
+    if (invalidate)
+        data.invalidate(USER);
 }
 
-function apiAsync(path: string, data?: any): Promise<any> {
+type ApiResult<T> = {
+    resp: T;
+    statusCode: number;
+    success: boolean;
+    errmsg: string;
+};
+
+function apiAsync<T = any>(path: string, data?: any): Promise<ApiResult<T>> {
     return U.requestAsync({
-        url: "/api/" + path,
-        headers: { "Cache-Control": "no-store" },
+        url: path,
+        headers: { "Cache-Control": "no-store" },   // don't cache responses
         method: data ? "POST" : "GET",
-        data: data || undefined
-    }).then(r => r.json).catch(core.handleNetworkError);
+        data: data || undefined,
+        withCredentials: true,  // include access token in request
```

Comment:
Don't we have APIs that don't require credentials?

Reply:
This just instructs the call to include cookies in cross-site requests (with matching origin, hence my need to change the CORS policy on the backend -- where I made that "CONFLICT" comment). Without this, the x-pxt-access-token cookie doesn't get sent.